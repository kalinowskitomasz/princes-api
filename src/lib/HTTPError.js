export default class HTTPError extends Error {
    constructor(message, status) {
        super(message)
        this.status = status
    }
}
