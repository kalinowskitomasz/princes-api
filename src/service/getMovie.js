import { StatusCodes } from 'http-status-codes'

import HTTPError from '../lib/HTTPError'

export default async function getMovie(id) {}
