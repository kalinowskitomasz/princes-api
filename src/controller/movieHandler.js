import { StatusCodes } from 'http-status-codes'

import HTTPError from '../lib/HTTPError'
import getMovie from '../service/getMovie'

// starts with tt prefix, then 7-10 digits
const imdbRegex = /^tt\d{7,10}$/
export function validateId(id) {
    if (typeof id !== 'string') return false

    return imdbRegex.test(id)
}

export default async function movieHandler({ id }, event) {
    if (!validateId(id))
        throw new HTTPError('Incorrect id', StatusCodes.BAD_REQUEST)

    const movieData = await getMovie(id)

    return new Response(JSON.stringify(movieData, null), {
        headers: { 'content-type': 'application/json;charset=UTF-8' },
    })
}
