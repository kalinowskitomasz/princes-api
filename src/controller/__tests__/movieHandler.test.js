import { StatusCodes } from 'http-status-codes'

import movieHandler, { validateId } from '../movieHandler'
import getMovie from '../../service/getMovie'
import HTTPError from '../../lib/HTTPError'

jest.mock('../../service/getMovie')

describe('validateId', () => {
    test('returns true on correct id', () => {
        expect(validateId('tt1234567')).toBeTruthy()
        expect(validateId('tt12345678')).toBeTruthy()
    })

    test('returns false on correct id', () => {
        expect(validateId('t1234567')).toBeFalsy()
        expect(validateId('aaaatt12345678')).toBeFalsy()
    })
})

describe('movieHandler', () => {
    beforeEach(() => {
        getMovie.mockReset()
    })

    test('returns bad request on incorrect id', async () => {
        await expect(movieHandler({ id: 'aatt1234567' }, null)).rejects.toEqual(
            new HTTPError('Incorrect id', StatusCodes.BAD_REQUEST)
        )
    })

    test('returns bad request on missing id', async () => {
        await expect(movieHandler({}, null)).rejects.toEqual(
            new HTTPError('Incorrect id', StatusCodes.BAD_REQUEST)
        )
    })

    test('returns data for correct id', async () => {
        getMovie.mockResolvedValue({
            ID: 'tt2527336',
            Title: 'Star Wars: Episode VIII - The Last Jedi',
            Year: '2017',
            Rated: 'PG-13',
            Released: '15 Dec 2017',
            Runtime: '152 min',
            Genre: 'Action, Adventure, Fantasy, Sci-Fi',
            Director: 'Rian Johnson',
            Writer:
                'Rian Johnson, George Lucas (based on characters created by)',
            Actors: 'Mark Hamill, Carrie Fisher, Adam Driver, Daisy Ridley',
            Plot:
                'Rey develops her newly discovered abilities with the guidance of Luke Skywalker, who is unsettled by the strength of her powers. Meanwhile, the Resistance prepares for battle with the First Order.',
            Language: 'English',
            Country: 'USA',
            Poster:
                'https://m.media-amazon.com/images/M/MV5BMjQ1MzcxNjg4N15BMl5BanBnXkFtZTgwNzgwMjY4MzI@._V1_SX300.jpg',
            Type: 'movie',
            Production: 'Walt Disney Pictures',
            Price: {
                'Film World': 24,
                'Cinema World': 24,
            },
        })
        await expect(
            movieHandler({ id: 'tt2527336' }, null)
        ).resolves.toMatchSnapshot()
    })
})
