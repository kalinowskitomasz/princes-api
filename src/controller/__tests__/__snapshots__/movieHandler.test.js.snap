// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movieHandler returns data for correct id 1`] = `
Response {
  "size": 0,
  "timeout": 0,
  Symbol(Body internals): Object {
    "body": Object {
      "data": Array [
        123,
        34,
        73,
        68,
        34,
        58,
        34,
        116,
        116,
        50,
        53,
        50,
        55,
        51,
        51,
        54,
        34,
        44,
        34,
        84,
        105,
        116,
        108,
        101,
        34,
        58,
        34,
        83,
        116,
        97,
        114,
        32,
        87,
        97,
        114,
        115,
        58,
        32,
        69,
        112,
        105,
        115,
        111,
        100,
        101,
        32,
        86,
        73,
        73,
        73,
        32,
        45,
        32,
        84,
        104,
        101,
        32,
        76,
        97,
        115,
        116,
        32,
        74,
        101,
        100,
        105,
        34,
        44,
        34,
        89,
        101,
        97,
        114,
        34,
        58,
        34,
        50,
        48,
        49,
        55,
        34,
        44,
        34,
        82,
        97,
        116,
        101,
        100,
        34,
        58,
        34,
        80,
        71,
        45,
        49,
        51,
        34,
        44,
        34,
        82,
        101,
        108,
        101,
        97,
        115,
        101,
        100,
        34,
        58,
        34,
        49,
        53,
        32,
        68,
        101,
        99,
        32,
        50,
        48,
        49,
        55,
        34,
        44,
        34,
        82,
        117,
        110,
        116,
        105,
        109,
        101,
        34,
        58,
        34,
        49,
        53,
        50,
        32,
        109,
        105,
        110,
        34,
        44,
        34,
        71,
        101,
        110,
        114,
        101,
        34,
        58,
        34,
        65,
        99,
        116,
        105,
        111,
        110,
        44,
        32,
        65,
        100,
        118,
        101,
        110,
        116,
        117,
        114,
        101,
        44,
        32,
        70,
        97,
        110,
        116,
        97,
        115,
        121,
        44,
        32,
        83,
        99,
        105,
        45,
        70,
        105,
        34,
        44,
        34,
        68,
        105,
        114,
        101,
        99,
        116,
        111,
        114,
        34,
        58,
        34,
        82,
        105,
        97,
        110,
        32,
        74,
        111,
        104,
        110,
        115,
        111,
        110,
        34,
        44,
        34,
        87,
        114,
        105,
        116,
        101,
        114,
        34,
        58,
        34,
        82,
        105,
        97,
        110,
        32,
        74,
        111,
        104,
        110,
        115,
        111,
        110,
        44,
        32,
        71,
        101,
        111,
        114,
        103,
        101,
        32,
        76,
        117,
        99,
        97,
        115,
        32,
        40,
        98,
        97,
        115,
        101,
        100,
        32,
        111,
        110,
        32,
        99,
        104,
        97,
        114,
        97,
        99,
        116,
        101,
        114,
        115,
        32,
        99,
        114,
        101,
        97,
        116,
        101,
        100,
        32,
        98,
        121,
        41,
        34,
        44,
        34,
        65,
        99,
        116,
        111,
        114,
        115,
        34,
        58,
        34,
        77,
        97,
        114,
        107,
        32,
        72,
        97,
        109,
        105,
        108,
        108,
        44,
        32,
        67,
        97,
        114,
        114,
        105,
        101,
        32,
        70,
        105,
        115,
        104,
        101,
        114,
        44,
        32,
        65,
        100,
        97,
        109,
        32,
        68,
        114,
        105,
        118,
        101,
        114,
        44,
        32,
        68,
        97,
        105,
        115,
        121,
        32,
        82,
        105,
        100,
        108,
        101,
        121,
        34,
        44,
        34,
        80,
        108,
        111,
        116,
        34,
        58,
        34,
        82,
        101,
        121,
        32,
        100,
        101,
        118,
        101,
        108,
        111,
        112,
        115,
        32,
        104,
        101,
        114,
        32,
        110,
        101,
        119,
        108,
        121,
        32,
        100,
        105,
        115,
        99,
        111,
        118,
        101,
        114,
        101,
        100,
        32,
        97,
        98,
        105,
        108,
        105,
        116,
        105,
        101,
        115,
        32,
        119,
        105,
        116,
        104,
        32,
        116,
        104,
        101,
        32,
        103,
        117,
        105,
        100,
        97,
        110,
        99,
        101,
        32,
        111,
        102,
        32,
        76,
        117,
        107,
        101,
        32,
        83,
        107,
        121,
        119,
        97,
        108,
        107,
        101,
        114,
        44,
        32,
        119,
        104,
        111,
        32,
        105,
        115,
        32,
        117,
        110,
        115,
        101,
        116,
        116,
        108,
        101,
        100,
        32,
        98,
        121,
        32,
        116,
        104,
        101,
        32,
        115,
        116,
        114,
        101,
        110,
        103,
        116,
        104,
        32,
        111,
        102,
        32,
        104,
        101,
        114,
        32,
        112,
        111,
        119,
        101,
        114,
        115,
        46,
        32,
        77,
        101,
        97,
        110,
        119,
        104,
        105,
        108,
        101,
        44,
        32,
        116,
        104,
        101,
        32,
        82,
        101,
        115,
        105,
        115,
        116,
        97,
        110,
        99,
        101,
        32,
        112,
        114,
        101,
        112,
        97,
        114,
        101,
        115,
        32,
        102,
        111,
        114,
        32,
        98,
        97,
        116,
        116,
        108,
        101,
        32,
        119,
        105,
        116,
        104,
        32,
        116,
        104,
        101,
        32,
        70,
        105,
        114,
        115,
        116,
        32,
        79,
        114,
        100,
        101,
        114,
        46,
        34,
        44,
        34,
        76,
        97,
        110,
        103,
        117,
        97,
        103,
        101,
        34,
        58,
        34,
        69,
        110,
        103,
        108,
        105,
        115,
        104,
        34,
        44,
        34,
        67,
        111,
        117,
        110,
        116,
        114,
        121,
        34,
        58,
        34,
        85,
        83,
        65,
        34,
        44,
        34,
        80,
        111,
        115,
        116,
        101,
        114,
        34,
        58,
        34,
        104,
        116,
        116,
        112,
        115,
        58,
        47,
        47,
        109,
        46,
        109,
        101,
        100,
        105,
        97,
        45,
        97,
        109,
        97,
        122,
        111,
        110,
        46,
        99,
        111,
        109,
        47,
        105,
        109,
        97,
        103,
        101,
        115,
        47,
        77,
        47,
        77,
        86,
        53,
        66,
        77,
        106,
        81,
        49,
        77,
        122,
        99,
        120,
        78,
        106,
        103,
        52,
        78,
        49,
        53,
        66,
        77,
        108,
        53,
        66,
        97,
        110,
        66,
        110,
        88,
        107,
        70,
        116,
        90,
        84,
        103,
        119,
        78,
        122,
        103,
        119,
        77,
        106,
        89,
        52,
        77,
        122,
        73,
        64,
        46,
        95,
        86,
        49,
        95,
        83,
        88,
        51,
        48,
        48,
        46,
        106,
        112,
        103,
        34,
        44,
        34,
        84,
        121,
        112,
        101,
        34,
        58,
        34,
        109,
        111,
        118,
        105,
        101,
        34,
        44,
        34,
        80,
        114,
        111,
        100,
        117,
        99,
        116,
        105,
        111,
        110,
        34,
        58,
        34,
        87,
        97,
        108,
        116,
        32,
        68,
        105,
        115,
        110,
        101,
        121,
        32,
        80,
        105,
        99,
        116,
        117,
        114,
        101,
        115,
        34,
        44,
        34,
        80,
        114,
        105,
        99,
        101,
        34,
        58,
        123,
        34,
        70,
        105,
        108,
        109,
        32,
        87,
        111,
        114,
        108,
        100,
        34,
        58,
        50,
        52,
        44,
        34,
        67,
        105,
        110,
        101,
        109,
        97,
        32,
        87,
        111,
        114,
        108,
        100,
        34,
        58,
        50,
        52,
        125,
        125,
      ],
      "type": "Buffer",
    },
    "disturbed": false,
    "error": null,
  },
  Symbol(Response internals): Object {
    "counter": undefined,
    "headers": Headers {
      Symbol(map): Object {
        "Access-Control-Allow-Origin": Array [
          "https://test",
        ],
        "content-type": Array [
          "application/json;charset=UTF-8",
        ],
      },
    },
    "status": 200,
    "statusText": "OK",
    "url": undefined,
  },
}
`;
